### PR TITLE
DataModule: common base class to reduce code duplication

### PR DESCRIPTION
**Pull request by adamjstewart**
A lot of the code in our base classes was shared, so I aggregated it.

Originally I was just planning on adding better error messages, but figured I would do this first.

### CODE REVIEW

File: torchgeo/datamodules/geo.py [hunk 1]
```diff
@@ -23,15 +23,76 @@
 from .utils import MisconfigurationException
 
 
-class GeoDataModule(LightningDataModule):  # type: ignore[misc]
-    """Base class for data modules containing geospatial information.
+class DataModule(LightningDataModule):  # type: ignore[misc]
```
Comment (calebrob6) [inline torchgeo/datamodules/geo.py:26]: Should we call this a TorchgeoDataModule or something?

**Review by calebrob6 — COMMENTED** [review]
This feels kind of like premature optimization (is premature abstracting a thing?) as it only saves a net few lines (just repeated twice), and "DataModule" doesn't seem like the correct name. It is good because I can't think of a case why we would want to go back and make different `plot`, `on_after_batch_transfer` or `prepare_data` methods for the Geo and NonGeo data modules.

**Comment by adamjstewart** [comment]
For context, the reason I made this change was because I wanted to improve the datamodule error messages. The current error message just says:
```
Foo.setup does not define a 'foo_dataset'
```
but the actual error could be any of:

* Foo.setup does not define a foo_dataset or dataset
* Foo.setup does not define a foo_sampler or sampler or foo_batch_sampler or batch_sampler
* The dataset length is 0
* The sampler length is 0

Since this is a bit of code and requires access to self, I figured it would make sense to add this check to a base class. When I started working on a base class, that's when I realized how much duplicated code there was. So I figured it would be easiest to add the base class in one PR and improve error messages in a second PR.

File: torchgeo/datamodules/geo.py (hunk 1 above)
Comment (adamjstewart) [reply torchgeo/datamodules/geo.py:26]: Could, I don't think it matters too much. The user should never use this class.

**Comment by calebrob6** [comment]
Totally fine! I just think the name "DataModule" might be confusing, maybe "BaseDataModule"? Else, I'm cool with it.

**Review by calebrob6 — APPROVED** [review]

**Pull request merged**